tests/unit/templates/test_jinja_scoping_lint.py

```
@@ -0,0 +1,153 @@
+_SET_RE = re.compile(r'{%-?\s*set\s+([a-zA-Z_][a-zA-Z0-9_]*)\s*=')
+_FOR_RE = re.compile(r'{%-?\s*for\s')
+_ENDFOR_RE = re.compile(r'{%-?\s*endfor')
+_STRING_RE = re.compile(r"'[^']*'|\"[^\"]*\"")
+
+
+def find_set_in_for_scoping_bugs(source):
+    """Return list of (line_number, variable) scoping violations in a template."""
+    source = _COMMENT_RE.sub(lambda m: ' ' * len(m.group(0)), source)
+    tags = [(m.start(), m.group(0)) for m in _TAG_RE.finditer(source)]
+
+    def line_of(pos):
```

Ruff (0.15.20)

[warning] Missing return type annotation for private function `line_of`

(ANN202)